### PR TITLE
JCSP-421: DC, Cloud: Switch between issues in search results

### PR DESCRIPTION
**Pull request by rsaliiatlassian**

### CODE REVIEW

File: custom-vu/src/main/kotlin/jces1209/vu/action/SwitchBetweenIssuesInSearchResultsCloud.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package jces1209.vu.action
+
+import com.atlassian.performance.tools.jiraactions.api.ActionType
+import com.atlassian.performance.tools.jiraactions.api.WebJira
+import com.atlassian.performance.tools.jiraactions.api.action.Action
+import com.atlassian.performance.tools.jiraactions.api.measure.ActionMeter
+import jces1209.vu.page.CloudIssueNavigator
+
+class SwitchBetweenIssuesInSearchResultsCloud(
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/action/SwitchBetweenIssuesInSearchResultsCloud.kt]: 1. It's to detailed for action. Please make like WorkOnSearch
2. SwitchBetweenIssuesInSearchResultsCloud and SwitchBetweenIssuesInSearchResultsDc are the same
Make one class but specify different pages for cloud and dc

**Review by ysemenkov — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt [hunk 2]
```diff
@@ -1,41 +1,66 @@
...
         .timeout(Duration.ofSeconds(30))
         .build()
 
     fun waitForNavigator() {
         page.waitForPageToLoad()
     }
+
+    fun chooseElement(): WebElement {
+        val elements = webDriver.wait(
+            ExpectedConditions.presenceOfAllElementsLocatedBy(By.xpath("//*[@class ='issue-list']/*"))
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt]:
```suggestion
            ExpectedConditions.presenceOfAllElementsLocatedBy(By.className("issue-list"))
```

File: custom-vu/src/main/kotlin/jces1209/vu/action/SwitchBetweenIssuesInSearchResultsCloud.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    private val meter: ActionMeter
+) : Action {
+    override fun run() {
+        jira.goToIssueNavigator("resolution = Unresolved ORDER BY priority DESC")
+        val cloudIssueNavigator = CloudIssueNavigator(jira.driver)
+        cloudIssueNavigator.waitForNavigator()
+        val element = cloudIssueNavigator.chooseElement()
+
+        meter.measure(
+            key = ActionType("Switch between issues in search results") { Unit },
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/action/SwitchBetweenIssuesInSearchResultsCloud.kt]: Create `ActionType` in MeasureType

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt [hunk 4]
```diff
@@ -1,41 +1,66 @@
...
+    )
         .cloudErrors()
         .timeout(Duration.ofSeconds(30))
         .build()
 
     fun waitForNavigator() {
         page.waitForPageToLoad()
     }
+
+    fun chooseElement(): WebElement {
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt]: It has the confusing name and it's redundant
you can get webelement in `switchBetweenIssuesInSearchResults` method or make it private and rename like 'getIssueElementFromList'

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt [hunk 5]
```diff
@@ -1,41 +1,66 @@
...
+        val elements = webDriver.wait(
+            ExpectedConditions.presenceOfAllElementsLocatedBy(By.xpath("//*[@class ='issue-list']/*"))
+        )
+        val rndIndex = Random().nextInt(elements.size - 1)
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt]: also, it's necessary to verify `is not 'class="focused"'`

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt [hunk 6]
```diff
@@ -1,41 +1,66 @@
...
+        return elements[rndIndex]
+    }
+
+    fun switchBetweenIssuesInSearchResults(element: WebElement) {
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt]: Please call it more shortly like `selectIssue` or `selectOtherIssue`

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt [hunk 7]
```diff
@@ -1,41 +1,66 @@
...
+        val dataKey = element.getAttribute("data-key")
+        val title = element.getAttribute("title")
+        webDriver.wait(
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt]: you just can make element.click()

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt [hunk 8]
```diff
@@ -1,41 +1,66 @@
...
 import org.openqa.selenium.support.ui.ExpectedConditions.and
 import org.openqa.selenium.support.ui.ExpectedConditions.or
 import org.openqa.selenium.support.ui.ExpectedConditions.presenceOfElementLocated
 import java.time.Duration
+import java.util.*
 
 class CloudIssueNavigator(
     driver: WebDriver
 ) {
+    private val webDriver = driver
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt]: it's redundant

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt [hunk 9]
```diff
@@ -1,41 +1,66 @@
...
+
+    fun switchBetweenIssuesInSearchResults(element: WebElement) {
+        val dataKey = element.getAttribute("data-key")
+        val title = element.getAttribute("title")
+        webDriver.wait(
+            ExpectedConditions.elementToBeClickable(By.xpath("//*[@data-key = '$dataKey']"))
+        )
+            .click()
+        webDriver.wait(
+            ExpectedConditions.presenceOfElementLocated(By.xpath("//*[@data-test-id = 'issue.views.issue-base.foundation.summary.heading' and contains(text(), '$title')]"))
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt]: Please use `performance profile` to see which the last elements are being loaded and wait for them
as i could see it's `comments` section, `watching` button and `menu`

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+class DcIssueNavigator(
+    driver: WebDriver
+) {
+    private val webDriver = driver
+    private val page = FalliblePage.Builder(
+        driver,
+        presenceOfElementLocated(By.className("issue-list"))
+
+    )
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt]:
```suggestion
    )
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.WebElement
+import org.openqa.selenium.support.ui.ExpectedConditions.*
+import java.time.Duration
+import java.util.*
+
+
+class DcIssueNavigator(
+    driver: WebDriver
+) {
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt]: It looks like there are a lot of the same things for DC and Cloud
Please extract them to the parent class and reuse

**Review by ysemenkov — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/action/SwitchBetweenIssuesInSearchResultsCloud.kt (hunk 1 above)
Comment (ysemenkov) [reply custom-vu/src/main/kotlin/jces1209/vu/action/SwitchBetweenIssuesInSearchResultsCloud.kt]: or even it's redundant, you can use SearchCloudFilter and SearchServerFilter actions

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt (hunk 2 above)
Comment (rsaliiatlassian) [reply custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt]: I'm receiving child elements //\*[@class ='issue-list']/\* . Won't change

File: custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt [hunk 12]
```diff
@@ -88,11 +90,17 @@ class ScenarioSimilarities(
...
+            jira = jira,
+            meter = meter
+        ),
         workOnSprint = WorkOnSprint(
             meter = meter,
             scrumBoardsMemory = boardsMemory.scrum,
             jiraTips = JiraTips(jira.driver)
-        )
+        ),
+        switchBetweenIssuesInSearchResults = switchBetweenIssuesInSearchResults
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt]: It looks like redundant

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt [hunk 13]
```diff
@@ -1,41 +1,70 @@
...
+        )
+        driver.wait(
+            ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@aria-label='Not watching']"))
+        )
+        driver.wait(
+            ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@data-test-id='issue.activity.comment']"))
+        )
+        driver.wait(
+            ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@aria-label='Add attachment']"))
+        )
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt]: use `ExpectedConditions.and`

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+class DcIssueNavigator(
+    driver: WebDriver
+): IssueNavigator {
+    private val webDriver = driver
+    private val page = FalliblePage.Builder(
+        driver,
+        presenceOfElementLocated(By.className("issue-list"))
+    )
+        .timeout(Duration.ofSeconds(30))
+        .build()
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt]:
```suggestion
    )
        .timeout(Duration.ofSeconds(30))
        .serverErrors()
        .build()
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+    override fun waitForNavigator() {
+        page.waitForPageToLoad()
+    }
+
+    override fun getIssueElementFromList(): WebElement {
+        val elements = webDriver.wait(
+            presenceOfAllElementsLocatedBy(By.xpath("//*[@class ='issue-list']/*"))
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt:44]: also, you could this locator extract to object field and reuse in other method - like FalliblePage
```suggestion
            presenceOfAllElementsLocatedBy(By.className("issue-list"))
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+import org.openqa.selenium.WebElement
+import org.openqa.selenium.support.ui.ExpectedConditions.*
+import java.time.Duration
+import java.util.*
+
+
+class DcIssueNavigator(
+    driver: WebDriver
+): IssueNavigator {
+    private val webDriver = driver
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt]: again, it's redundant - just use `driver`

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        )
+        return webDriver.wait(
+            elementToBeClickable(elementLocator)
+        )
+    }
+
+    override fun selectIssue(element: WebElement) {
+        val dataKey = element.getAttribute("data-key")
+        element.click()
+        webDriver.wait(
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt]: It waits for low number of elements. Did you check by chrome profiler which last elements are being loaded?

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+        val elementLocator = By.xpath("//*[@data-key = '$dataKey']")
+        webDriver.wait(
+            visibilityOfElementLocated(elementLocator)
+        )
+        webDriver.wait(
+            presenceOfAllElementsLocatedBy(elementLocator)
+        )
+        return webDriver.wait(
+            elementToBeClickable(elementLocator)
+        )
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt]: it looks like enough to wait for `elementToBeClickable`

File: custom-vu/src/main/kotlin/jces1209/vu/page/IssueNavigator.kt [hunk 19]
```diff
@@ -0,0 +1,9 @@
+package jces1209.vu.page
+
+import org.openqa.selenium.WebElement
+
+interface IssueNavigator {
+    fun waitForNavigator()
+    fun getIssueElementFromList(): WebElement
+    fun selectIssue(element: WebElement)
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/IssueNavigator.kt]: What about to encapsulate work on selenium objects and merge methods to one?
`getIssueElementFromList()` + `selectIssue()` = `selectOtherIssue()`

**Review by ysemenkov — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt (hunk 15 above)
Comment (rsaliiatlassian) [reply custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt:44]: Here I'm getting child elements of //*[@class ='issue-list'] element. So left as is

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt (hunk 18 above)
Comment (rsaliiatlassian) [reply custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt]: It doesn't work with just elementToBeClickable(elementLocator) require additional checks

File: custom-vu/src/main/kotlin/jces1209/vu/page/IssueNavigator.kt (hunk 19 above)
Comment (rsaliiatlassian) [reply custom-vu/src/main/kotlin/jces1209/vu/page/IssueNavigator.kt]: In WorkOnSearchResults action code firstly check that all is loaded and select some element with which going to work. After that in measure section works with that element

val element = issueNavigator.getIssueElementFromList()

        meter.measure(
            key = SWITCH_BETWEEN_ISSUES_IN_SEARCH_RESULTS,
            action = {
                issueNavigator.selectIssue(element)
            }
        )

File: custom-vu/src/main/kotlin/jces1209/vu/JiraCloudScenario.kt [hunk 20]
```diff
@@ -64,7 +65,8 @@ class JiraCloudScenario : Scenario {
                 jira = jira,
                 meter = meter,
                 columnsEditor = CloudColumnsEditor(jira.driver)
-            )
+            ),
+            issueNavigator = CloudIssueNavigator(driver = jira.driver)
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/JiraCloudScenario.kt]:
```suggestion
            issueNavigator = CloudIssueNavigator(jira.driver)
```

File: custom-vu/src/main/kotlin/jces1209/vu/JiraDcScenario.kt [hunk 21]
```diff
@@ -64,7 +65,8 @@ class JiraDcScenario : Scenario {
                 jira = jira,
                 meter = meter,
                 columnsEditor = DcColumnsEditor(jira.driver)
-            )
+            ),
+            issueNavigator = DcIssueNavigator(driver = jira.driver)
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/JiraDcScenario.kt]:
```suggestion
            issueNavigator = DcIssueNavigator(jira.driver)
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/IssueNavigator.kt (hunk 19 above)
Comment (ysemenkov) [reply custom-vu/src/main/kotlin/jces1209/vu/page/IssueNavigator.kt]: I don't suggest to merge `issueNavigator.waitForNavigator()`
Selecting issue could be a part of the switching issue - it should be no problem to measure it
so the client of page object doesn't work with WebElement objects

**Review by ysemenkov — APPROVED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt [hunk 22]
```diff
@@ -1,41 +1,67 @@
...
+            )
+        )
+    }
+
+    private fun getIssueElementFromList(): WebElement {
+        val elements = driver.wait(
+            ExpectedConditions.presenceOfAllElementsLocatedBy(By.xpath("//*[@class ='issue-list']/*[not(@class ='focused')]"))
+        )
+        val rndIndex = Random().nextInt(elements.size - 1)
+        return elements[rndIndex]
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt:65]: We should use seeded randoms to get repeatable tests. So gotta pass the `SeededRandom` it from `Scenario` to `CloudIssueNavigator` via constructor and use it:
```kotlin
random.pick(elements)!!
```

**Pull request merged**